src/NFT.sol

```diff
@@ -0,0 +1,191 @@
+     */
+    modifier whenMintingIsActive() {
+        require(mintingActive, "Minting is not active");
+        _;
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                              CONSTRUCTOR
+    //////////////////////////////////////////////////////////////*/
+    uint256 public immutable maxSupply; // Maximum issuance of NFTs. 0 means unlimited.
+    uint256 public nextTokenId = 1; // Next token id to be minted id's are 1 based
+    string public baseURI;
+    bool public metadataFrozen;
+    address public minter;
+    bool mintingActive = true;
+
+    /**
+     * @notice Creates a new instance of NFT
+     * @param _name Name of the NFT
+     * @param _symbol Symbol of the NFT
+     * @param _uri Base URI of the NFT, concatenated with Token ID to create tokenURI
+     * @param _minter Address of the minter
+     * @param _maxSupply Maximum supply of NFTs. 0 means unlimited
+     */
+    constructor(
+        string memory _name,
+        string memory _symbol,
+        string memory _uri,
+        address _minter,
+        uint256 _maxSupply
+    ) ERC721(_name, _symbol) {
+        _setBaseURI(_uri);
+        _setMinter(_minter);
+        maxSupply = _maxSupply;
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                                METADATA
+    //////////////////////////////////////////////////////////////*/
+
+    /**
+    @dev Updates the baseURI.
+    @param newBaseURI New Base URI Value
+    */
+    function _setBaseURI(string memory newBaseURI) internal onlyOwner {
+        if (metadataFrozen) {
+            revert METADATA_IS_IMMUTABLE();
+        }
+        baseURI = newBaseURI;
+        emit URIChanged(baseURI);
+    }
+
+    /**
+    @dev Freezes the metadata uri.
+    */
+    function freezeMetadataURI() external onlyOwner {
+        metadataFrozen = true;
+        emit MetadataFrozen();
+    }
+
+    function tokenURI(uint256 tokenId)
+        public
+        view
+        virtual
+        override
+        returns (string memory)
+    {
+        return string(abi.encodePacked(baseURI, "/", tokenId));
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                                MINTING
+    //////////////////////////////////////////////////////////////*/
+
+    /**
+    @dev Ends minting permanently.
+    */
+    function endMinting() external onlyOwner whenMintingIsActive {
+        mintingActive = false;
+    }
+
+    /**
+    @dev Allows composable minting contracts to check if minting is active.
+    @return bool True if minting is active.
+    */
+    function isMintingActive() external view returns (bool) {
+        return mintingActive;
+    }
+
+    /**
+    @dev Checks if max. supply has been reached.
+    @return bool Flag indicating if max. supply has been reached.
+    */
+    function isMaxSupplyReached() public view returns (bool) {
+        if (maxSupply > 0 && nextTokenId == maxSupply) {
```

Comment:
i think the `maxSupply > 0` is going to cause a problem in `_mint` for collections with unlimited supply